processor/src/advice/dsa.rs

```diff
@@ -0,0 +1,104 @@
+use super::{ExecutionError, Felt, StarkField, Vec, Word};
+use vm_core::crypto::dsa::rpo_falcon512::{
+    KeyPair, Polynomial, PublicKeyBytes, SecretKeyBytes, PK_LEN, SK_LEN,
+};
+
+/// Gets as input a vector containing an expanded public key and its associated secret key, and a
+/// word representing a message and outputs a vector of values to be pushed onto the advice stack.
+/// The values are the ones required for a Falcon signature verification inside the VM and they are:
+///
+/// 1. The nonce represented as 8 field elements.
+/// 2. The expanded public key represented as the coefficients of a polynomial of degree < 512.
+/// 3. The signature represented as the coefficients of a polynomial of degree < 512.
+/// 4. The product of the above two polynomials in the ring of polynomials with coefficients
+/// in the Miden field.
+///
+/// Inputs:
+///   Operand stack: [PK, MSG, ...]
+///   Advice stack: [...]
+///   Advice map: {PK: [pk_raw, sk_raw]}
+///
+/// Outputs:
+///   Operand stack: [PK, MSG, ...]
+///   Advice stack: [NONCE1, NONCE0, h, s2, pi]
+///   Advice map: {PK: [pk_raw, sk_raw]}
+///
+/// Where:
+/// - PK is the digest of an expanded public.
+/// - MSG is the digest of the message to be signed.
+/// - [NONCE0, NONCE1] is a double-word representing a 40 bit nonce that is used in the Falcon
+/// hash-to-point algorithm.
+/// - h is the polynomial representing the expanded public key corresponding to the digest PK.
+/// - s2 is the polynomial representing the signature with the secret key associated to PK on
+/// the message MSG.
+/// - pi is the product of the above two polynomials.
+/// - pk_raw are raw bytes of the expanded public key.
+/// - sk_raw are raw bytes of the secret key.
+///
+/// # Errors
+/// Will return an error if either:
+/// - The advice map does not contain an entry with key PK.
+/// - The advice map entry under key PK is not a vector of the expected length.
+/// ///
+/// The function only generates non-deterministic input that is required for the Falcon verification
+/// procedure inside the VM and as such does interact with the VM only through the advice provider.
+pub fn falcon_sign(pk_sk: &[Felt], msg: Word) -> Result<Vec<Felt>, ExecutionError> {
+    if pk_sk.len() != (PK_LEN + SK_LEN) {
+        return Err(ExecutionError::AdviceStackReadFailed(0));
+    }
+
+    // To generate a signature, we need the expanded key as well as the secret key
+    let pk_exp = pk_sk[..PK_LEN].to_vec();
+    let pk_exp: PublicKeyBytes = vec_felt_to_u8(&pk_exp)
+        .try_into()
+        .expect("Should not fail as we've checked the length of the combined vector");
+
+    let sk = pk_sk[PK_LEN..].to_vec();
+    let sk: SecretKeyBytes = vec_felt_to_u8(&sk)
+        .try_into()
+        .expect("Should not fail as we've checked the length of the combined vector");
+
+    // Create the corresponding key pair
+    let key_pair = KeyPair::new_existing(&pk_exp, &sk).expect("Should not fail given the above");
```

Comment:
Another way to do this is something like this:
```Rust
let mut key_pair_bytes = Vec::with_capacity(pk_sk.len());
for element in pk_sk {
    let value = element.as_int();
    if value > u8::MAX as u64 {
        return Err(ExecutionError::MalformedSignatureKey("RPO Falcon512"));
    }
    key_pair_bytes.push(value as u8);
}

let key_pair = KeyPair::read_from_bytes(&key_pair_bytes)
    .map_err(|_| ExecutionError::MalformedSignatureKey("RPO Falcon512"))?;
```
This would require importing `Deserializable` from `vm_core::utils`, but we'd avoid making any changes in `miden-crypto` crate.

Reply:
Makes sense!